base/scopedvalues.jl

```diff
@@ -0,0 +1,120 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module ScopedValues
+
+export ScopedValue, scoped
+
+"""
+    ScopedValue(x)
+
+Create a container that propagates values across dynamic scopes.
+Use [`scoped`](@ref) to create and enter a new dynamic scope.
+
+Values can only be set when entering a new dynamic scope,
+and the value referred to will be constant during the
+execution of a dynamic scope.
+
+Dynamic scopes are propagated across tasks.
+
+# Examples
+```jldoctest
+julia> const svar = ScopedValue(1);
+
+julia> svar[]
+1
+
+julia> scoped(svar => 2) do
+           svar[]
+       end
+2
+```
+
+!!! compat "Julia 1.11"
+    Scoped values were introduced in Julia 1.11. In Julia 1.7+ a compatible
+    implementation is available from the package ScopedValues.jl.
+"""
+mutable struct ScopedValue{T}
+    const initial_value::T
+    ScopedValue{T}(initial_value) where {T} = new{T}(initial_value)
+end
+ScopedValue(initial_value::T) where {T} = ScopedValue{T}(initial_value)
+
+Base.eltype(::Type{ScopedValue{T}}) where {T} = T
+
+import Base: ImmutableDict
+
+##
+# Notes on the implementation.
+# We want lookup to be unreasonable fast.
+# - IDDict/Dict are ~10ns
+# - ImmutableDict is faster up to about ~15 entries
+# - ScopedValue are meant to be constant, Immutabilty
+#   is thus a boon
+# - If we were to use IDDict/Dict we would need to split
+#   the cache portion and the value portion of the hash-table,
+#   the value portion is read-only/write-once, but the cache version
+#   would need a lock which makes ImmutableDict incredibly attractive.
+#   We could also use task-local-storage, but that added about 12ns.
+# - Values are GC'd when scopes become unreachable, one could use
+#   a WeakKeyDict to also ensure that values get GC'd when ScopedValues
+#   become unreachable.
+# - Scopes are an inline implementation of an ImmutableDict, if we wanted
+#   be really fance we could use a CTrie or HAMT.
+
+mutable struct Scope{T}
+    const parent::Union{Nothing, Scope}
+    const key::ScopedValue{T}
+    const value::T
+end
+Scope(parent, key::ScopedValue{T}, value) where T =
+    Scope(parent, key, convert(T, value))
+
+current_scope() = current_task().scope::Union{Nothing, Scope}
+
+function Base.show(io::IO, ::Scope)
+    print(io, Scope)
+end
```

Comment:
why do we define this? it seems like all it does is hide potentially useful info, but then again i don't know in what context we'd be printing `Scope`s?

Reply:
Yeah, I was debating if I should print the actual values.

But it would just be `ScopedValue(1) => 1` and then eventually `ScopedValue(1) => 2` xD.
ScopedValue's are not named.

We could print `objectid(ScopedValue) => 1`, but the complexity didn't seem super worthwhile.

Reply:
x-ref: https://github.com/vchuravy/ScopedValues.jl/pull/5

Reply:
yeah, i think that's nice. I can only imagine inspecting scopes for debugging and i think this would be helpful in that case 👍